### PR TITLE
KAFKA-9939; Fix overcounting delayed fetches in request rate metrics

### PR DESCRIPTION
**Pull request by hachikuji**
Fetches which hit purgatory are currently counted twice in fetch request rate metrics. This patch moves the metric update into `fetchMessages` so that they are only counted once. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by hachikuji** [comment]
The failure in the test case is a problem with the isolation of metrics between test cases. I will submit a fix shortly.

File: core/src/test/scala/unit/kafka/server/SimpleFetchTest.scala [hunk 1]
```diff
@@ -1,206 +0,0 @@
...
-   *
-   * When a fetch operation with read committed data turned on is received, the replica manager
-   * should only return data up to the HW of the partition; when a fetch operation with read
-   * committed data turned off is received, the replica manager could return data up to the LEO
-   * of the local leader replica's log.
-   *
-   * This test also verifies counts of fetch requests recorded by the ReplicaManager
-   */
-  @Test
-  def testReadFromLog(): Unit = {
```
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/server/SimpleFetchTest.scala:171]: I decided to get rid of this. I believe it is already covered by test cases in `ReplicaManagerTest`. See for example, `testFetchBeyondHighWatermarkReturnEmptyResponse`.

**Comment by edoardocomar** [comment]
Hi @hachikuji as you're working on this ... do you mind taking a look at https://github.com/apache/kafka/pull/4204 ?

the test we have added there still fails with this change of yours, without the workarounds we had suggested. 
Thanks

**Comment by edoardocomar** [comment]
cc @mimaison

File: core/src/test/scala/unit/kafka/server/SimpleFetchTest.scala (hunk 1 above)
Comment (ijuma) [reply core/src/test/scala/unit/kafka/server/SimpleFetchTest.scala:171]: That test name is a bit misleading. Should we name it `testFetchBeyondHighwatermarkForConsumerAndFollower` or something?

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 2]
```diff
@@ -1017,6 +1018,51 @@ class ReplicaManagerTest {
...
+    assertNotNull(nonPurgatoryFetchResult.get)
+    assertEquals(Errors.NONE, nonPurgatoryFetchResult.get.error)
+    assertMetricCount(1)
+
+    val purgatoryFetchResult = sendConsumerFetch(replicaManager, tp0, partitionData, None, timeout = 10)
+    assertNull(purgatoryFetchResult.get)
+    mockTimer.advanceClock(11)
+    assertNotNull(purgatoryFetchResult.get)
+    assertEquals(Errors.NONE, purgatoryFetchResult.get.error)
+    assertMetricCount(2)
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:1065]: This would return `3` without the fix?

**Review by ijuma — APPROVED** [review]
Nice catch, LGTM. A question and a minor comment below.

**Comment by hachikuji** [comment]
@edoardocomar Thanks for the comment. I agree it is related. Left a comment on #4204.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala (hunk 2 above)
Comment (hachikuji) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:1065]: Right.

File: core/src/test/scala/unit/kafka/server/SimpleFetchTest.scala (hunk 1 above)
Comment (hachikuji) [reply core/src/test/scala/unit/kafka/server/SimpleFetchTest.scala:171]: Yeah, that's fair. How about just `testFetchBeyondHighwatermark`?

Comment (ijuma) [reply core/src/test/scala/unit/kafka/server/SimpleFetchTest.scala:171]: That's fine. Is it worth adding a brief comment stating what it's testing?

Comment (hachikuji) [reply core/src/test/scala/unit/kafka/server/SimpleFetchTest.scala:171]: Borderline overkill I guess, but I added a few comments explaining the test behavior.

**Review by ijuma — APPROVED** [review]
LGTM

**Comment by hachikuji** [comment]
The 32 test failures on jdk11 were due to threads being left behind after a failure in `TransactionsBounceTest`. I will submit a separate fix for this. The other failure seems to be one of the recently flaky streams tests.

**Pull request merged**